src/Components/src/Microsoft.AspNetCore.Components/Rendering/HtmlRenderer.cs

```diff
@@ -0,0 +1,234 @@
+        private int RenderElement(
+            List<string> result,
+            ArrayRange<RenderTreeFrame> frames,
+            int position)
+        {
+            ref var frame = ref frames.Array[position];
+            result.Add("<");
+            result.Add(frame.ElementName);
+            var afterAttributes = RenderAttributes(result, frames, position + 1, frame.ElementSubtreeLength - 1);
+            var remainingElements = frame.ElementSubtreeLength + position - afterAttributes;
+            if (remainingElements > 0)
+            {
+                result.Add(">");
+                var afterElement = RenderChildren(result, frames, afterAttributes, remainingElements);
+                result.Add("</");
+                result.Add(frame.ElementName);
+                result.Add(">");
+                Debug.Assert(afterElement == position + frame.ElementSubtreeLength);
+                return afterElement;
+            }
+            else
+            {
+                if (SelfClosingElements.Contains(frame.ElementName))
+                {
+                    result.Add(" />");
+                }
+                else
+                {
+                    result.Add(">");
+                    result.Add("</");
+                    result.Add(frame.ElementName);
+                    result.Add(">");
+                }
+                Debug.Assert(afterAttributes == position + frame.ElementSubtreeLength);
+                return afterAttributes;
+            }
+        }
+
+        private int RenderChildren(List<string> result, ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return position;
+            }
+
+            return RenderFrames(result, frames, position, maxElements);
+        }
+
+        private int RenderAttributes(
+            List<string> result,
+            ArrayRange<RenderTreeFrame> frames, int position, int maxElements)
+        {
+            if (maxElements == 0)
+            {
+                return position;
+            }
+
+            for (var i = 0; i < maxElements; i++)
+            {
+                var candidateIndex = position + i;
+                ref var frame = ref frames.Array[candidateIndex];
+                if (frame.FrameType != RenderTreeFrameType.Attribute)
+                {
+                    return candidateIndex;
+                }
+
+                if (frame.AttributeValue is bool flag && flag)
+                {
+                    result.Add(" ");
+                    result.Add(frame.AttributeName);
+                }
+
+                if (frame.AttributeValue is string value)
+                {
+                    result.Add(" ");
+                    result.Add(frame.AttributeName);
+                    result.Add("=");
+                    result.Add("\"");
+                    result.Add(_htmlEncoder.Encode(value));
+                    result.Add("\"");
+                }
```

Comment:
If we know it's a bool but is false, we don't also need to check if it's a string. Suggest changing the flow to:

```csharp
                if (frame.AttributeValue is bool flag)
                {
                    if (flag)
                    {
                        result.Add(" ");
                        result.Add(frame.AttributeName);
                    }
                }
                else if (frame.AttributeValue is string value)
                {
                    result.Add(" ");
                    result.Add(frame.AttributeName);
                    result.Add("=");
                    result.Add("\"");
                    result.Add(_htmlEncoder.Encode(value));
                    result.Add("\"");
                }
```

Reply:
Changed to:
```csharp
                switch (frame.AttributeValue)
                {
                    case bool flag when flag:
                        result.Add(" ");
                        result.Add(frame.AttributeName);
                        break;
                    case string value:
                        result.Add(" ");
                        result.Add(frame.AttributeName);
                        result.Add("=");
                        result.Add("\"");
                        result.Add(_htmlEncoder(value));
                        result.Add("\"");
                        break;
                    default:
                        break;
                }
```